### PR TITLE
VMware: New module vmware_host_hyperthreading

### PR DESCRIPTION
**Pull request by ckotte**
##### SUMMARY
Enable or disable Hyperthreading optimization for an ESXi host.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vmware_host_hyperthreading

##### ANSIBLE VERSION
```
ansible 2.6.3
  config file = /root/ansible-vmware/ansible.cfg
  configured module search path = [u'/root/ansible-vmware/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### ADDITIONAL INFORMATION

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: s/ESXI/ESXi/

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: Missing period at end.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: s/vmware_host_powermgmt_policy/vmware_host_hyperthreading/

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: s/vmware_host_powermgmt_policy/vmware_host_hyperthreading/

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: s/vmware_host_powermgmt_policy/vmware_host_hyperthreading/

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: Add additional information about previous and current state of hyperthreading so that user can do stuff depending on that information in next tasks of playbook.

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (Akasurde) [inline test/integration/targets/vmware_host_hyperthreading/tasks/main.yml]: s/vmware_host_acceptance/vmware_host_hyperthreading/

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (Akasurde) [inline test/integration/targets/vmware_host_hyperthreading/tasks/main.yml]: Add your name here.

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (Akasurde) [inline test/integration/targets/vmware_host_hyperthreading/tasks/main.yml]: s/Acceptance Level/Hyperthreading/

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (Akasurde) [inline test/integration/targets/vmware_host_hyperthreading/tasks/main.yml:91]: Add testcases for Cluster with and without check_mode.

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: @Akasurde Can you give me an example? I think you implemented this already in one of your other modules!? Otherwise, I would do something like this:
```
results['result'][host.name]['before'] = ...
results['result'][host.name]['after'] = ...
```

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (ckotte) [reply test/integration/targets/vmware_host_hyperthreading/tasks/main.yml]: Sorry, I copied the tests from one of your modules.

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (ckotte) [reply test/integration/targets/vmware_host_hyperthreading/tasks/main.yml]: Sorry, I copied the tests from one of your modules.

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (ckotte) [reply test/integration/targets/vmware_host_hyperthreading/tasks/main.yml:91]: I cannot enable Hyperthreading on ESXi hosts in vcsim anyway, but I added cluster tests as well.

File: test/integration/targets/vmware_host_hyperthreading/tasks/main.yml
Comment (ckotte) [reply test/integration/targets/vmware_host_hyperthreading/tasks/main.yml]: done

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: Something similar - https://github.com/ansible/ansible/blob/fb72a5424cab097d28670d00cf107fd515cee3f0/lib/ansible/modules/cloud/vmware/vmware_host_lockdown.py#L114

Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: @Akasurde What's the benefit of `current_state`. Looks like it's similar to `desired_state` all the time!?

Then you can access the state like this?:
```
- name: Enable Hyperthreading
  tags: hyperthreading
  vmware_host_hyperthreading:
    hostname: "{{ vcenter_server }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
    esxi_hostname: "{{ inventory_hostname }}"
    enabled: true
    validate_certs: false
  register: hyperthreading_state
  delegate_to: "{{ jump_server }}"

- name: Hyperthreading status
  tags: hyperthreading
  debug:
    var: hyperthreading_state['result']['{{ inventory_hostname }}']['current_state']
```

Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: @Akasurde Could you please answer my question? I need that information for other modules as well. Thanks.

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: `current_state` is state after performing vmware task and we can use that information in next ansible task in playbook.
```yaml
  register: hyperthreading_state
  delegate_to: "{{ jump_server }}"

- name: Perform some task depending upon current state of hyperthreading
  debug:
    var: "I am here"
  when: hyperthreading_state['result']['{{ inventory_hostname }}']['current_state'] == 'enabled'
```

Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: And use `desired_state` only for reference? They have similar values (if we don't have any errors, but the play would fail anyway). They would be only different if you run in check mode, but this depends on how it's implemented.

Enable Hyperthreading (no check mode)
```
results['result'][host.name]['desired_state'] = "enabled"
results['result'][host.name]['previous_state'] = "disabled"
results['result'][host.name]['current_state'] = "enabled"
```
Enable Hyperthreading (check mode)
```
results['result'][host.name]['desired_state'] = "enabled"
results['result'][host.name]['previous_state'] = "disabled"
results['result'][host.name]['current_state'] = "disabled"
```

Does this makes sense?

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: I think in check mode is like dry-run without any action. so Enable hyperthreading in check mode should have current_state and desired_state as enabled.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: isn't it better practice to do `and not hyperthreading_info.active` instead?

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: It would be good to call out a reboot is required to make it active

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: Same as with the other module: I suggest setting the current/desired state to what it would actually become, including the correct message. Check mode should represent what will be the outcome of the change as if it were for real, except not actually make the change.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: Same remark on best practice for `xxx is False` vs `not xxx`, unless you anticipate that it can be another value than True or False?

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: Call out that a reboot is needed to actually deactivate it.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: previous state in this case would be `enabled`,

additionally on the message: Same as with the other module: I suggest setting the current/desired state to what it would actually become, including the correct message. Check mode should represent what will be the outcome of the change as if it were for real, except not actually make the change.

**Review by pdellaert — CHANGES_REQUESTED** [review]
Overall good module, some remarks on check mode and cleanup, but minor stuff :)

**Comment by adarobin** [comment]
This is cool and all, but why not just use:
```
- name: Disable Hyperthreading
  vmware_host_config_manager:
    hostname: '{{ vcenter_hostname }}'
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
    esxi_hostname: '{{ inventory_hostname }}'
    options:
        'VMkernel.Boot.Hyperthreading': false
  delegate_to: localhost
```
?

**Comment by ckotte** [comment]
I didn't know that setting. Thanks for the hint, but how do you want to make sure that Hyperthreading is also active or really disabled?

For example, Hyperthreading is enabled on all of our 300+ hosts, but I found a few hosts with this module where it was not active. It was enabled in ESXi, but for whatever reason, it was disabled in the BIOS (which was basically a mismatch in the desired state ESXi host config I want to implement/ensure).

Later, I've put the hosts in maintenance mode, configured the BIOS setting, and rebooted the server. Everything with Ansible.

**Comment by adarobin** [comment]
@ckotte Sorry, it was the end of the day and I guess I wasn't looking at the code closely enough to realize you were enabling hyperthreading _and_ checking the current status of it.

I ran this against a few of my test hosts and there is a corner case that probably needs to be documented.

If you are setting the `VMkernel.Boot.hyperthreadingMitigation` advanced option to `true` to [mitigate against L1TF/Foreshadow](https://kb.vmware.com/s/article/55806) and you are setting the state of hyperthreading to enabled - You will always get the "Hyperthreading is already enabled for host '%s', but not active. A reboot is required!" message and rebooting will not make a difference.

**Comment by ckotte** [comment]
@adarobin No problem. Maybe `host.config.hyperThread.active` is not set or not available if you set this advanced setting. 

Do you use `VMkernel.Boot.hyperthreadingMitigation` and have Hyperthreading enabled in your environment? If so, I assume you don't see some info in the vSphere client? You can see "Reboot needed" or something similar that usually correlates to the value of `host.config.hyperThread.active`.

I will test that option later this week. Thanks for checking the module.

**Comment by adarobin** [comment]
@ckotte We are rolling out `VMkernel.Boot.hyperthreadingMitigation` across our hosts and we do have Hyperthreading currently enabled in our environment.  It seems to make the checkbox in the vSphere client not work per this [VMware Communities thread](https://communities.vmware.com/thread/597236).

`VMkernel.Boot.hyperthreadingMitigation` basically disables Hyperthreading when the OS is booting on hardware that is vulnerable without disabling it globally.  The thought behind this is that if they make improvements to their scheduler that allow for the use of Hyperthreading safely on vulnerable hardware in the future (i.e. [Hyper-V's HyperClear](https://blogs.technet.microsoft.com/virtualization/2018/08/14/hyper-v-hyperclear/)), you just need to install the patch and reboot and Hyperthreading will start to be in use again.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: Yes, makes sense. I've changed it. Thanks.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: No, based on my testings, it can only be True or False. I've changed it. Thanks.

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: I've added the reboot information to the output. Thanks.

**Comment by ckotte** [comment]
@adarobin I tested `VMkernel.Boot.hyperthreadingMitigation` with an old HPE G8 blade. The GUI shows that I need a reboot and the module says that Hyperthreading isn't active; also after a reboot. 
```
TASK [esxi : Enable Hyperthreading] **********************************************************************************************************************************************************************************************************
changed: [esxi_host -> jump_server] => {"changed": true, "result": {"esxi_host": {"current_state": "enabled", "desired_state": "enabled", "error": "", "msg": "Hyperthreading is enabled for host 'esxi_host', but not active. A reboot is required!", "previous_state": "enabled"}}}
```

However, Hyperthreading isn't actually active. The amount of Logical Processors in the GUI is correct, but when I check the CPUs with `esxtop`, I can only see half of it.

Can you confirm? 

So, basically the module works as expected. Either leave it like it is and update the description or the module could check if the advanced setting is set and change the message to something like `"Hyperthreading is enabled for host 'esxi_host', but not active to mitigate against L1TF/Foreshadow ('VMkernel.Boot.hyperthreadingMitigation' is set to 'true').` or something like that?

P.S.: I couldn't set `VMkernel.Boot.hyperthreadingMitigation` with `vmware_host_config_manager`. I got the error `Unknown option ...` all the time. I could change the setting only after setting it manually the first time. Did you change the setting with Ansible on your hosts or manually?

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py]: I did that too. Thanks.

**Comment by adarobin** [comment]
@ckotte [VMware KB 55806 ](https://kb.vmware.com/s/article/55806) states "Note: Enabling this option will result in the vSphere UI reporting only a single logical processor per physical core; halving the number of logical processors if Hyperthreading was previously enabled. In addition Hyperthreading may be reported as 'Disabled' in various configuration tabs."  The check you propose would work, but just documenting this is probably fine.  The main thing is that if I use this module as is, all of my hosts with the mitigation enabled will always report changed.

Also, the reason you can't set it is because the option was only created after certain patch levels. 
This is the task I am using to set it on my hosts:
```
- name: Enable Foreshadow mitigation on hosts that support it
  vmware_host_config_manager:
    hostname: '{{ vcenter_hostname }}'
    esxi_hostname: '{{ inventory_hostname }}'
    options:
        'VMkernel.Boot.hyperthreadingMitigation': true
  delegate_to: localhost
  when: (ansible_distribution_version == "6.5.0" and ansible_distribution_build|int >= 9298722) or (ansible_distribution_version == "6.0.0" and ansible_distribution_build|int >= 9313334)
```

**Comment by ckotte** [comment]
> @ckotte [VMware KB 55806 ](https://kb.vmware.com/s/article/55806) states "Note: Enabling this option will result in the vSphere UI reporting only a single logical processor per physical core; halving the number of logical processors if Hyperthreading was previously enabled. In addition Hyperthreading may be reported as 'Disabled' in various configuration tabs." 

The thing is. The GUI doesn't show me a single logical processor per physical core. It's not reported as disabled on my side, but it is actually disabled. Looks like this implementation is still buggy.

> The check you propose would work, but just documenting this is probably fine. The main thing is that if I use this module as is, all of my hosts with the mitigation enabled will always report changed.

I will add a workaround that changed won't be reported in this case. We might need to enable this as well if we upgrade our ESXi hosts.

> Also, the reason you can't set it is because the option was only created after certain patch levels.

I updated the host because I couldn't find the setting. After that I still couldn't set it via Ansible. Need to investigate with other hosts.

> This is the task I am using to set it on my hosts:
> 
> ```
> - name: Enable Foreshadow mitigation on hosts that support it
>   vmware_host_config_manager:
>     hostname: '{{ vcenter_hostname }}'
>     esxi_hostname: '{{ inventory_hostname }}'
>     options:
>         'VMkernel.Boot.hyperthreadingMitigation': true
>   delegate_to: localhost
>   when: (ansible_distribution_version == "6.5.0" and ansible_distribution_build|int >= 9298722) or (ansible_distribution_version == "6.0.0" and ansible_distribution_build|int >= 9313334)
> ```

Thanks. I will add that to my playbook as well.

**Comment by adarobin** [comment]
> I will add a workaround that changed won't be reported in this case. We might need to enable this as well if we upgrade our ESXi hosts.

Sounds great!

> I updated the host because I couldn't find the setting. After that I still couldn't set it via Ansible. Need to investigate with other hosts.

What version of Ansible are you using?  I was having weird issues with `vmware_host_config_manager` on 2.6.x.  For example, I could set `NFS.MaxVolumes` on ESXi 6.5 hosts but not 6.0 hosts.  Updating to 2.7.0 which has #41343 seems to have resolved the issue.

**Comment by ckotte** [comment]
I've added the L1 Terminal Fault (L1TF)/Foreshadow mitigation workaround. Now, the output gives the user a hint that Hyperthreading is inactive because the processor is vulnerable:
```
TASK [esxi : Enable Hyperthreading] **********************************************************************************************************************************************************************************************************
ok: [esxi01 -> jump_server] => {
  "changed": false,
  "result": {
    "esxi01": {
      "current_state": "enabled",
      "desired_state": "enabled",
      "msg": "Hyperthreading is enabled, but not active because the processor is vulnerable to L1 Terminal Fault (L1TF).",
      "previous_state": "enabled"
    }
  }
}
```

File: lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_host_hyperthreading.py:95]: Done. This just doesn't work if another task depends on `current_state` and would throw an error if Hyperthreading is not active.

**Comment by Akasurde** [comment]
@ckotte Are we Ok on merging this ? Code LGTM, let me know once I will merge this.

**Comment by ckotte** [comment]
> @ckotte Are we Ok on merging this ? Code LGTM, let me know once I will merge this.

Yes, please approve. We can discuss the check mode message format, standard parameters, the parameter naming, the examples format (https://github.com/ansible/ansible/pull/47271#discussion_r226405450 and https://github.com/ansible/ansible/issues/43433), etc. pp. in the WG and change all modules later together if necessary.

**Comment by ckotte** [comment]
> @ckotte Are we Ok on merging this ? Code LGTM, let me know once I will merge this.

I also added your remaining changes. I don't have access to my personal mail when connected to company VPN. I also use `git push --force` every time...

**Comment by ckotte** [comment]
@Akasurde Could you review this module again, please? 

It's now ready. I just changed the behavior of `desired`, `previous`, and `current` to match with my other PRs. Basically, `state_desired` is just `state` and `previous` will only be used if something has changed. `state_current` is used because we also have the states `active` and `inactive`.

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
@ckotte Thanks for the contribution.

**Pull request merged**
